lib/locale_threads.t

```diff
@@ -54,67 +471,1290 @@ SKIP: { # perl #127708
+                    || ! setlocale($map_category_name_to_number{$category_name},
+                                   $locale_name))
+                {
+                    die "Unexpectedly can't set locale to $locale_name from "
+                      . setlocale($LC_ALL)
+                      . "; \$!=$!, \$^E=$^E";
+                }
+            }
+
+            my $got = eval $eval_string;
+            next if $got eq $result
+                 && utf8::is_utf8($got) == utf8::is_utf8($result);
+
+            # If the result varied from the expected value, this is a
+            # non-deterministic locale, so, don't test it.
+            diag("For '$eval_string',\nresults in iteration $i differed from"
+               . " the original\ngot");
+            Dump($got);
+            diag("expected");
+            Dump($result);
+            next LOCALE;
+        }
+
+        # Here, the setlocale for this locale appears deterministic.  Use it.
+        my $op_result = pack_op_result($op, $result);
+        push $distincts{$category_name}{$op_result}{locales}->@*, $locale_name;
+        # No point in looking beyond this if we already have all the tests we
+        # need.  Note this assumes that the same op isn't used in two
+        # categories.
+        if (defined $op_counts{$op} && $op_counts{$op} >= $thread_count)
+        {
+            last;
+        }
+    }
+}
+
+use Config;
+
+# Figure out from config how to represent disparate LC_ALL
+my @valid_category_numbers = sort { $a <=> $b }
+                    map { $map_category_name_to_number{$_} } @valid_categories;
+
+my $use_name_value_pairs = defined $Config{d_perl_lc_all_uses_name_value_pairs};
+my $lc_all_separator = ($use_name_value_pairs)
+                       ? ";"
+                       : $Config{perl_lc_all_separator} =~ s/"//gr;
+my @position_to_category_number;
+if (! $use_name_value_pairs) {
+    my $positions = $Config{perl_lc_all_category_positions_init} =~ s/[{}]//gr;
+    $positions =~ s/,//g;
+    $positions =~ s/^ +//;
+    $positions =~ s/ +$//;
+    @position_to_category_number = split / \s+ /x, $positions
+}
+
+sub get_next_category() {
+    use feature 'state';
+    state $index;
+
+    # Called to rotate all the legal locale categories
+
+    my $which = ($use_name_value_pairs)
+                ? \@valid_category_numbers
+                : \@position_to_category_number;
+
+    $index = -1 unless defined $index;
+    $index++;
+
+    if (! defined $which->[$index]) {
+        undef $index;
+        return;
+    }
+
+    my $category_number = $which->[$index];
+    return $category_number if $category_number != $LC_ALL;
+
+    # If this was LC_ALL, the next one won't be
+    return &get_next_category();
+}
+
 SKIP: {
-    skip("POSIX version doesn't support thread-safe locale operations", 1)
-                                                unless ${^SAFE_LOCALES};
+    skip("Unsafe locale threads", 1) unless ${^SAFE_LOCALES};
```

Comment:
Can you add a comment in here as to how ${^SAFE_LOCALES} gets determined ("set") on a given machine?  (I can't figure that out from existing documentation in `perllocale` and `perlvar`.  Thanks.

Reply:
perlvar says

> ${^SAFE_LOCALES}
             Reflects if safe locale operations are available to this perl
             (when the value is 1) or not (the value is 0). This variable is
             always 1 if the perl has been compiled without threads. It is
             also 1 if this perl is using thread-safe locale operations. Note
             that an individual thread may choose to use the global locale
             (generally unsafe) by calling "switch_to_global_locale" in
             perlapi. This variable currently is still set to 1 in such
             threads.
  
 This makes sense to me but patches or suggestions are welcome.  If using locales is safe to do on this platform without fear of another thread coming along and screwing this thread up, then the value is 1.  You shouldn't be using locales on platforms where the value is 0, as, you can get segfaults, etc.

Reply:
github is not allowing me to reply to some of the separate comments individually, so here is a conglomeration.

I don't know how to shorten the failure output.  What I did was to write the got vs expected out and then used a diff tool that has per-character granularity.  This particular test is failing in the output of strxfrm() which is not designed to be human-readable.  What it shows is either the perl code is not doing the right thing on Darwin, or the Darwin libc has a race.  I think its the latter, as the code runs fine on OpenBSD.  This  test file usually passes on Darwin, with the occasional flap.  I've never seen strxfrm() fail on it.  I have looked at the libc code in Darwin before, and it is based on early *BSD code, that hasn't been synced with newer BSD releases; those releases have fixed bugs in this area, and continue to do so.  What this means is that Darwin cannot be considered to have thread-safe locales.  I'll turn off their use in another commit and repush this PR

Reply:
FreeBSD 13.1 does not have safe threaded locales, as  the failure in #21469 indicates.  Hence the test is skipped.

Reply:
Would it then be fair and accurate to say, *"Whether or not a particular machine has safe threaded locales is a characteristic of the machine over which the person compiling, building or using Perl has no control whatsoever?"*

Of course, that begs the question, *"How does Perl discover or calculate what the value of ${^SAFE_LOCALES} is for a particular machine?"*

Reply:
Not quite.  You have no control over whether a particular machine is capable of having safe threaded locales, except indirectly by filing bug reports when the machine wrongly claims such capability, and waiting if and until it gets fixed.

But you can Configure perl to not use the safe interface even if it is available.  I see no good reason to do so unless you discover that the safe interface is buggier than the unsafe one.

Perl knows that ${^SAFE_LOCALES} is true of non-threaded perls, as there is no other thread than your own.  Otherwise, Configure probes if the POSIX safe interface is available, and assumes it works as advertised.  But experience has shown  that this is a lie in many cases, so hints files are used to override this for particular versions.  Perl knows when Microsoft introduced safe threaded locales, and uses preprocessor conditionals to act on that.  Similarly, Perl knows that MingW needs to have been compiled with a modern libc dll, and checks for that

Reply:
I used this test to investigate failures, deciding that either it was perl's fault and fixed that, or that it was the platform's fault.  Usually in the latter case, I or others built a C reproducer program not involving perl.  So the answer to how the answer was discovered is that it it was mostly a manual process resulting in bug fixes to perl, or to hints files changes